arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/Option.kt

```diff
@@ -796,10 +925,14 @@ public sealed class Option<out A> {
    *   none.exists { it > 10 }      // Result: false
    * }
    * ```
-   * <!--- KNIT example-option-21.kt -->
+   * <!--- KNIT example-option-24.kt -->
    *
    * @param predicate the predicate to test
    */
+  @Deprecated(
```

Comment:
Exists also aligns with stdlib for list. I vote for keeping this.

Reply:
It's an API that was deprecated for `Either` after community review on the PR, https://github.com/arrow-kt/arrow/pull/2830.

We can consider keeping it, but it seemed like a rare case. In that case we should revert the deprecation on `Either`.

Reply:
Exists for either might not make much sense but for option I'd argue that it actually does. Essentially you can draw parallel of an optional value being a list of at most one element or otherwise none. If this api is removed then one would argue there's a missing functionality to assert that an option has a value with a certain quality. Lists for instance has `list.any { }` and `list.all { }`. If we're deprecating exists simply because it's deprecated in either, I would argue that we should perhaps reintroduce something that has a same function with a different name.

Reply:
> If we're deprecating exists simply because it's deprecated in either, I would argue that we should perhaps reintroduce something that has the same function with a different name.

I was actually thinking about this @myuwono, and then apply the same for `Either` but I couldn't think of a better API (or name). 😂 My main thought was that I also wanted something for in the DSL, but couldn't think of anything...

There is now `isLeft()` * `isRight()` with smart-contracts. We could also expose `isRight { }` and `isLeft { }` 🤔
Similarly we could do the same for `Option`, but they're currently named `isEmpty` and `isNotEmpty`.

How would you feel about changing `isEmpty` and `isNotEmpty` to `isNone()`/`isSome()`? And adding/replacing `exists` with `isNone { }` and `isSome { }`? Do you think that would result in a nicer API for `Option`(, and `Either`)?

It does cover more cases, whilst still keeping the API more aligned I think. They're also very in line with `onSome { }`/`onNone { }` and `onLeft { }`/`onRight { }`.

cc\\ @franciscodr

Reply:
Thanks for your very constructive feedback btw! 🙏 It's really valuable.

Reply:
I'm also up to rename `isEmpty` and `isNotEmpty` to `isNone` and `isSome`, respectively, to align them with Either API.

Regarding `isNone { }` and `isSome { }`, what would those functions do? Evaluate the block only if the emptiness/non-emptiness predicate holds?

Reply:
Oh, `isNone { }` doesn't make sense 😅 Something like this.

```kotlin
   public inline fun <A> Option<A>.isSome(predicate: (A) -> Boolean): Boolean {
    contract {
      callsInPlace(predicate, EXACTLY_ONCE)
      returns(true) implies (this@Option is Some<A>)
      returns(false) implies (this@Option is None)
    }
    return this is Some<A> && predicate(value)
  }
```

And for `Either` alongside `isLeft()` and `isRight`:

```kotlin
  public inline fun <A, B> Either<A, B>.isLeft(predicate: (A) -> Boolean): Boolean {
    contract {
      callsInPlace(predicate, EXACTLY_ONCE)
      returns(true) implies (this@Either is Left<A>)
      returns(false) implies (this@Either is Right<B>)
    }
    return this@Either is Left<A> && predicate(value)
  }

  public inline fun <A, B> Either<A, B>.isRight(predicate: (B) -> Boolean): Boolean {
    contract {
      callsInPlace(predicate, EXACTLY_ONCE)
      returns(true) implies (this@Either is Right<B>)
      returns(false) implies (this@Either is Left<A>)
    }
    return this@Either is Right<B> && predicate(value)
  }
```

WDYT @franciscodr @myuwono

Reply:
That's a really great idea!

Reply:
I like your proposal, @nomisRev. I sometimes had to double-check if `exists` returns `true`  when the option value was equal to `None` or on the contrary. Having a function called `isSome` make it more evident that the function will only return `true` if the option contains a value and that value satisfies the predicate.